DPI Improvements for wxTextCtrl, wxSearchCtrl, wxButton

MaartenBent opened the pull request:
Fixed height of wxTextCtrl too high after DPI increase.
Fixed height rich wxTextCtrl too small after DPI increase.
Improve positioning of text control inside wxSearchCtrl after DPI increase.
Let standard buttons on wxMSW have at least the height of wxBU_EXACTFIT.

See https://trac.wxwidgets.org/ticket/18563 and https://trac.wxwidgets.org/ticket/18551

MaartenBent committed:
Remove FromDIP in wxGetEditHeightFromCharHeight

This caused wxTextCtrl to become to high at higher DPI.

It was added to fix appearance of wxSearchCtrl at high DPI. Remove an unneeded
FromDIP there too, so it gets the same height as a normal wxTextCtrl.
And centre the textctrl inside the searchctrl.

MaartenBent committed:
Simplify wxSearchCtrl::DoGetBestClientSize()

MaartenBent committed:
Use 'fit exactly' on button page in widgets sample to toggle wxBU_EXACTFIT

MaartenBent committed:
Make button and searchctrl pages in widgets sample more DPI aware

MaartenBent committed:
Let standard buttons on wxMSW have at least the height of wxBU_EXACTFIT

Closes #18563

vadz commented on src/msw/anybutton.cpp:
```diff
@@ -437,16 +437,16 @@ wxSize wxMSWButton::IncreaseToStdSizeAndCache(wxControl *btn, const wxSize& size
-    }
+    // wxBU_EXACTFIT is typically used alongside a text control or similar,
+    // so make them as high as it.
+    // The standard hight is generally higher than this, but if not (e.g. when
```
```suggestion
    // The standard height is generally higher than this, but if not (e.g. when
```

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2538,8 +2538,18 @@ wxSize wxTextCtrl::DoGetBestSize() const
+
+    if ( IsRich() )
+    {
+        // Font if rich control is not adjusted to the active DPI, see
```
```suggestion
        // Font in rich control is not adjusted to the active DPI, see
```

vadz commented on include/wx/msw/private.h:
```diff
@@ -91,12 +91,13 @@ WXDLLIMPEXP_BASE void wxSetInstance(HINSTANCE hInst);
 //
-// The wxWindow parameter must be valid and used for getting the DPI.
-inline int wxGetEditHeightFromCharHeight(int cy, const wxWindow* w)
+// The wxWindow parameter is not used.
```
```suggestion
// The wxWindow parameter is currently not used but should still be valid.
```
Just in case we decided to use it again in the future.

vadz reviewed (approved):
Thanks, no real comments, just a couple of typos in the comments that I could also fix myself when merging if you prefer.

MaartenBent commented:
If there are no other comments (@PBfordev ?), you can fix them with merging. Thanks.

PBfordev commented:
My primary monitor is scaled at 100% (120 DPI), while the secondary one is at 100% and I can still see the issue with `wxTextCtrl` with `wxTE_RICH(2)` flag.

How to reproduce:
1. Run the widgets sample and move its window to the unscaled monitor.
2. Go to Native/Text page and in "Control kind" select either "rich edit" or "rich edit 2.0".
3. The font in the "Text:" becomes noticeably larger then for the plain edit. The size is corrected after the window is moved to the scaled screen and back.

When rich edits are created on the scaled screen, their font size matches the plain edit one.

MaartenBent commented:
> I can still see the issue with `wxTextCtrl` with `wxTE_RICH(2)` flag

I think I have fixed it. Can you try again. Thanks.

PBfordev commented:
Thanks for the quick reply. I have tested the updated PR on the same setup as in my previous post, the issue described there is gone.

Unfortunately, there is still an issue. Now the font size in rich edit controls created on the unscaled screen shrinks when moved to the scaled screen. The font remains tiny after that, regardless of the window being moved between the screens.

MaartenBent commented:
I removed DPI handling case for rich, I hope this fixes it. I could reproduce 'something' but I'm not sure if this was the issue you are referring to.

PBfordev commented:
Once again, thanks for the quick response, I now cannot reproduce either of the two issues I have described in this PR . However, I am afraid that I have to report another issue.

I have used also the **text** sample to test this PR (this sample always opens on the scaled screen and ATM I could not make it open on my unscaled screen so there may be some other issues). 

When moving the sample's window from the scaled screen to the unscaled one, the customized text colours (e.g., red, blue, and green at the beginning of the first paragraph) and the underlines in the right-most text control (the tall one) disappear. The background colour remains. I could not reproduce this with the master. 

However, the master already manifest another issue present also with this PR: when the sample window appears, the tall edit is scrolled to the bottom; when the window is moved to the unscaled screen, it scrolls to the top.

PBfordev commented:
Thanks again for the quick fix, now I cannot observe any issue I previously reported here, except for the scrollbar reset when moving the window from the scaled screen to the unscaled one.

But me being me, I have tried another test case (probably also testing your patience here as well): Changing the font in the rich edit. The text sample has a simple rich text editor which can be invoked via menu File / Show Rich Text Editor. When you change the font of a part of the text in this rich text editor (via menu Edit/Font), the font is reset when moving the window between the scaled and unscaled screen, regardless of the direction.

vadz commented on include/wx/msw/textctrl.h:
```diff
@@ -132,6 +132,7 @@ class WXDLLIMPEXP_CORE wxTextCtrl : public wxTextCtrlBase
     virtual bool SetBackgroundColour(const wxColour& colour) wxOVERRIDE;
     virtual bool SetForegroundColour(const wxColour& colour) wxOVERRIDE;
     virtual bool SetFont(const wxFont& font) wxOVERRIDE;
+    wxFont GetFont() const;
```
Sorry, this doesn't seem like a good idea: either we need to make `GetFont()` virtual in the base class or rename this to something like `MSWGetScaledFont()` to avoid hiding the base class method.

Ideal would probably be to update `m_font`, so that `GetFont()` returns the font of the same size as the one actually used by the control, but keep using unscaled `HFONT` for the native control.

MaartenBent replied:
I'll revert it to the original fix (using `WXAdjustToPPI` in `DoGetSizeFromTextSize`).

vadz commented on include/wx/msw/textctrl.h:
```diff
@@ -265,6 +266,10 @@ class WXDLLIMPEXP_CORE wxTextCtrl : public wxTextCtrlBase
     // can't nor really need to distinguish between them and 4 is for 4.1)
     int m_verRichEdit;
+
+    // rich text controls behave strange on DPI changes, keep track of the DPI
```
```suggestion
    // rich text controls behave strangely on DPI changes, keep track of the DPI
```

vadz reviewed (commented):
I don't think hiding `GetFont()` is a good idea, so I'd like to find another solution.

If you'd like, I could apply the other commits and leave this one for later.

PBfordev commented:
I may be missing something but to implement a simple solution adjusting font size in rich edit may not be possible. Rich edit:
1. Can have its parts formatted using many different fonts[1].
2. Does not provide a (n efficient) way to retrieve information about them.

Maybe doing something as in the [MS dynamic DPI sample](https://github.com/microsoft/Windows-classic-samples/blob/master/Samples/DynamicDPI/cpp/SampleRichEditControl.h) will be necessary, with internally adjusting the font passed in `{Get|Set}Font()` and `{Get|Set}(Default)Style()`. `EM_{GET|SET}ZOOM` being available only since Win10 1809 and user-zooming may be complications but perhaps it is the only working solution...

[1] Referring to the issue described in the second paragraph of [my comment here](https://github.com/wxWidgets/wxWidgets/pull/1634#issuecomment-549420814).

MaartenBent commented:
I found a solution by iterating over the text and updating the font, this is only necessary on the first DPI change and only when the richt text control is created on a display with non-system DPI.
I'll update (probably) later today.

MaartenBent committed:
Fix height of rich wxTextCtrl after DPI change

The font of rich controls is not adjusted when the DPI changes, so do it here.

MaartenBent commented:
This solution works for me. It could probably be optimized, for example using only one float with the scaling factor instead of two `wxSize` for the DPI.

Can you create an issue on trac for the srollbar problem. I haven't seen it and will not address it in this PR.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -3341,7 +3389,17 @@ bool wxTextCtrl::GetStyle(long position, wxTextAttr& style)
         lf.lfWeight = FW_NORMAL;
 
+    // Determine the pointSize that was used in SetStyle. Don't worry about
+    // lfHeight or PPI, style.SetFont() will loose this information anyway.
```
```suggestion
    // lfHeight or PPI, style.SetFont() will lose this information anyway.
```

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -339,6 +339,8 @@ void wxTextCtrl::Init()
 #if wxUSE_RICHEDIT
     m_verRichEdit = 0;
+    m_richBaseDPI = wxDefaultSize;
+    m_richUsedDPI = wxDefaultSize;
```
Seems unnecessary, i.e. it looks like we could just leave them default-initialized, but harmless, of course.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -517,6 +519,10 @@ bool wxTextCtrl::MSWCreateText(const wxString& value,
+
+        WindowHDC hdc(GetHWND());
+        m_richBaseDPI = wxSize(::GetDeviceCaps(hdc, LOGPIXELSX), ::GetDeviceCaps(hdc, LOGPIXELSY));
+        m_richUsedDPI = GetDPI();
```
Sorry, could you please explain how could this return a different value? `GetDPI()` uses `WindowHDC` with the `HWND` of the top-level parent, surely it should use the same DPI as `m_richBaseDPI`? What am I missing?

MaartenBent replied:
Imagine you have two displays. The main display with DPI 100% and a second display with DPI 150%. The system DPI will be 100% (96).
If you initialize the control on the second display, `m_richBaseDPI` will get value 96, and `m_richUsedDPI` will get the DPI of the second display (144).

vadz replied:
Wait, I see it now: you're using `WindowHDC(GetHWND())` here, but `m_hWnd` is still 0 because the window hadn't been created yet, so it's always the screen `HDC` for the primary screen in fact.

This is very, very confusing however (especially so when looking at the diff, which doesn't show that `MSWCreateControl()` is only called below) and I don't understand why would we do it like this instead of just using `ScreenHDC` directly.

Could you please change it like this or am I still missing something?

MaartenBent replied:
I think you're still missing it. `GetDeviceCaps` will always return the system DPI, not matter on what display it is.` GetDPI` will use `GetDpiForWindow` and return the DPI of the display it (or its parent) is on.

MaartenBent replied:
But using `GetHWND` here might not be the best thing to do. I see if I can move it to a better place.

vadz replied:
OK, thanks, I indeed forgot that `::GetDeviceCaps()` always used the main display values. But, still, if we need the main screen DPI, using `ScreenHDC` here would IMHO be more clear.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+    // change. Reapply the styles to the text.
+    else if ( m_richBaseDPI != m_richUsedDPI )
+    {
+        wxEventBlocker block(this, wxEVT_TEXT);
```
Hmm, does `SetStyle()` generate events?? I had no idea and it's probably wrong...

MaartenBent replied:
Yes, see a similar note at `wxTextCtrl::SetFont`.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+    {
+        wxEventBlocker block(this, wxEVT_TEXT);
+
+        wxSize baseDPI = m_richBaseDPI;
```
```suggestion
        const wxSize baseDPI = m_richBaseDPI;
```
Just to make it more clear what's going on here.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+
+            if ( rangeStart >= 0 || i == rangeTotal )
+            {
+                // update the baseDPI so no font size change is applied
```
```suggestion
                // update the base DPI so no font size change is applied
```
To avoid confusion with the local `baseDPI` variable.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+            rangeStart = i;
+            commonAttr = attr;
+        }
+    }
```
What's the value of `m_richBaseDPI` at the end here? It looks like it could have changed, shouldn't we restore it to its original value?

MaartenBent replied:
It ends with the value of `m_richUsedDPI`. So any subsequent DPI changes, this code is not executed anymore.

vadz replied:
OK, thanks, I see now that we always execute `SetStyle()` call due to `i == rangeTotal` check. This is a bit tortuous though, I think doing
```cpp
m_richBaseDPI = baseDPI;
GetStyle(i, attr);
m_richBaseDPI = m_richUsedDPI;
```
and not touching `m_richBaseDPI` anywhere else would make it more clear that we just use the base DPI temporarily for `GetStyle()` call duration and that the base DPI always is the same as used DPI at the end.

vadz reviewed (commented):
I must admit that I'm rather confused by this and don't even understand how can used/base DPI values turn out to be different at all. But if it works, let's merge this nevertheless, of course.

I'd just like to have some conceptual idea of how `RICHEDIT` deals with DPI: does it always use current DPI for scaling when rendering? I.e. does it use `used/base` DPI scale factor internally?

MaartenBent commented:
> I'd just like to have some conceptual idea of how `RICHEDIT` deals with DPI: does it always use current DPI for scaling when rendering? I.e. does it use `used/base` DPI scale factor internally?

Basically, we don't have to do anything. Except in one case: when the rich text control is created on a screen with a different DPI than the system DPI. In this case we have to apply a scale factor to the font. After the control changes DPI, this scaling has to be reset and all works fine again.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+        wxSize baseDPI = m_richBaseDPI;
+
+        wxTextAttr attr;
+        wxTextAttr commonAttr;
```
BTW, couldn't `attr == commonAttr` below be written just as `attr.IsDefault()`? I think this would be more clear (and slightly more efficient).

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+            if ( attr == commonAttr && i != rangeTotal )
+                continue;
+
+            if ( rangeStart >= 0 || i == rangeTotal )
```
After staring at these conditions for 5 minutes, I still don't find them as obvious as I'd like. Could we write this loop like this instead:
```cpp
// Store the old base DPI value, it will be changed to the actually used
// new DPI below and will be left with this value after the end of this
// function, so we won't need to do anything during the next DPI change.
const wxSize baseDPI = m_richBaseDPI;

const int rangeTotal = GetLastPosition();
for ( int rangeStart = -1, i = 0; ; ++i )
{
    // Temporarily set base DPI to get the actually used font size.
    m_richBaseDPI = baseDPI;
    GetStyle(i, attr);
    m_richBaseDPI = m_richUsedDPI;

    if ( i == rangeTotal )
    {
        // Always set the attribute for the last span of text, we need
        // to do it if only to update its font size.
        SetStyle(rangeStart, i, commonAttr);
        break;
    }

    if ( attr == commonAttr )
    {
        // Still using the same attribute, don't change rangeStart below.
        continue;
    }

    // The attribute has changed, so apply the previous attribute to the
    // range it was used for, if non-empty.
    if ( rangeStart >= 0 )
    {
        SetStyle(rangeStart, i, commonAttr);
    }

    // And remember the new attribute, as well as the start of the range
    // it applies to.
    rangeStart = i;
    commonAttr = attr;
}
```

Personally I find this more clear, what do you think?

MaartenBent replied:
I updated the loop to look like this.

vadz commented:
> > I'd just like to have some conceptual idea of how `RICHEDIT` deals with DPI: does it always use current DPI for scaling when rendering? I.e. does it use `used/base` DPI scale factor internally?
> 
> Basically, we don't have to do anything. Except in one case: when the rich text control is created on a screen with a different DPI than the system DPI. In this case we have to apply a scale factor to the font. After the control changes DPI, this scaling has to be reset and all works fine again.

This really looks like a bug in RICHEDIT itself and I wonder what's going to happen if they fix it in a next update. But I guess we have no choice but to wait and see...

The only alternative I see is to always (artificially) create the control as a child of the desktop (i.e. always on the primary display) and then reparent it under its real parent. This would definitely be ugly but OTOH it could end up being simpler than this code and should continue to work even if RICHEDIT changes in the future. Of course, I haven't even tested whether this actually works...

PBfordev commented:
I am afraid I have to confirm what I hinted at in my previous post. Iterating over all characters in the text control (each iteration probably results in two window messages sent?) results, at least on my PC, in a quite bad user experience.

How I tested:
1. Launch the text sample, move its window to the unscaled monitor, select Show Rich Text Editor.
2. Attempt to drag the Rich Text Editor from the unscaled to the scaled monitor.
3. Observe the window get stuck between the monitors and the application become unresponsive. 
4. After a couple of seconds, I can move the Rich Text Editor window back to the unscaled monitor (but not to the monitor where I was dragging it to) but the it is unresponsive, displaying a tiny font. 
5. After another about 10 seconds the application becomes usable again.

I have AMD's Ryzen 5 2600 which is not a slow CPU; the editor has mere 11,800 characters, I would assume the duration of computation would proportionally increase with the text length. 

I tested 64-bit DLL Release and Debug configurations (built with CMake and MSVC 2019).

MaartenBent commented:
That's unfortunate. I'll try using `EM_SETZOOM` then.

PBfordev commented:
Mhm, I have run the Dynamic DPI sample I referred above and it seem that its UI is exactly the same size on both my monitors. It is unexpected as normally the size of the native UI elements in centimeters on the unscaled monitor is somewhat smaller (22" 1650x1080 at 96 DPI vs 26" 2560x1440 at 120 DPI).

vadz commented:
@PBfordev thanks for testing! I thought about this, but decided it couldn't be _that_ bad... except that it clearly is. We could postpone updating the fonts and do it during idle time, but I'm afraid it might be not much better if the control lazily continued to redraw itself in the new font 10 seconds after finishing dragging it to another screen.

The hack with always creating these controls on the main display starts looking more appealing...

@MaartenBent maybe I should merge the other commits and leave the rich text ones for another PR?

MaartenBent commented:
> The hack with always creating these controls on the main display starts looking more appealing...

That might not be possible if there is only one display of which the DPI is changed while the application is running.

> @MaartenBent maybe I should merge the other commits and leave the rich text ones for another PR?

I'm almost done with using `EM_SETZOOM`.

vadz commented:
> 
> 
> > The hack with always creating these controls on the main display starts looking more appealing...
> 
> That might not be possible if there is only one display of which the DPI is changed while the application is running.

I think the control must be using something like `::GetDeviceCaps(LOGPIXELSz)` internally, so I'd hope that it still uses the primary display DPI even if it's different from 96. Do you think it just hard codes 96 instead?

Of course, if you have a working zoom-based implementation, it's irrelevant anyhow.

> > @MaartenBent maybe I should merge the other commits and leave the rich text ones for another PR?
> 
> I'm almost done with using `EM_SETZOOM`.

Great, thanks again for all your work on this!

PBfordev commented:
Thanks again for the quick response and of course, all the work. 

I ran out of time for today so I could do only a brief testing. 

The only issue I have found is that when the Rich Text Editor window in the text sample is created on the unscaled screen, the font is too large  and remains large even after moving between the unscaled and scaled screen and back. It is odd as all the other rich edits in the text and widgets samples look OK when created on the unscaled screen.

I have few more tests in mind but no time left, but I will try to do them tomorrow evening (UTC+1).

MaartenBent commented:
Looks like the Rich Text Editor window problem is related to creating and setting text in the RichTextFrame constructor. If I use the code below scaling is applied correctly, but c++11 is not yet allowed. Maybe setting the initial zoom somewhere in `wxTextCtrl::MSWHandleMessage` works.
```
CallAfter([=] {
    ApplyZoomFactor(GetHWND(), m_richDPIscale);
});
```

> The documentation says `EM_SETZOOM ` is supported only since Win10 v1809

The defines itself have been available since Rich Edit 3.0. They added `ES_EX_ZOOMABLE` later.

MaartenBent committed:
Fix font size of rich wxTextCtrl when created on display with non-system DPI

Apply a zoom factor based on the active DPI and the system DPI. On the first
DPI change, revert the scaling factor.

PBfordev commented:
Thanks. I could not find any issue besides that reported in my previous comment.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -339,6 +339,7 @@ void wxTextCtrl::Init()
 {
 #if wxUSE_RICHEDIT
     m_verRichEdit = 0;
+    m_richDPIscale = 0;
```
I am not sure if this should be 0 or 1?

MaartenBent replied:
I set it to 1 now, so I could remove a not-zero check.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -608,6 +609,10 @@ bool wxTextCtrl::MSWCreateText(const wxString& value,
+
+        // Zoom is not applied when this control is created in a wxFrame
+        // constructor, so delay applying it.
+        CallAfter(&wxTextCtrl::MSWSetInitialZoom);
```
Not a big deal, but perhaps we should initialize `m_richDPIscale` to its real value here and only call `CallAfter()` if it's different from 1?

MaartenBent replied:
I added a `== 1` check to `MSWSetRichZoom` so zoom is only set when it is required.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2718,14 +2725,53 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+inline void ApplyZoomFactor(HWND hwnd, float scale)
+{
+    UINT num = 1;
+    UINT denom = 1;
```
It looks like we could just define them as `[WL]PARAM` and avoid some casts.

MaartenBent replied:
`EM_GETZOOM` uses them by reference, I don't think that will work.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2718,14 +2725,53 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+        ratio = ratio * (num / (float)denom);
+
+    num = 100 * ratio;
+    denom = 100;
```
Why do we multiply by 100? It looks like multiplying by 12 or 24 or maybe 60 would be better as it would result in integer results for all the common scale factors.

MaartenBent replied:
Windows seems to use 100 as default for denominator, at least when using ctrl+mousewheel to zoom. I added a comment in the code.

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -2718,14 +2725,53 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
+    // Determine the system DPI and the DPI of the display the rich control is
+    // shown on, and calculate and apply the scaling factor.
+    m_richDPIscale = GetDPI().y / (float)::GetDeviceCaps(ScreenHDC(), LOGPIXELSY);
+    ApplyZoomFactor(GetHWND(), m_richDPIscale);
```
I think we should skip this call if `m_richDPIscale == 1.0f`.

vadz reviewed (approved):
Thanks, I think this can be merged now -- unless you plan any more changes?

vadz commented:
Thanks again, Maarten!

I've tested the latest version in my usual configuration (Windows 10 v1803, 2 displays, one with 100% and another one with 200% DPI) and things mostly look good, although, annoyingly, there are still some slight discrepancies between the widgets sample and the native stuff (I checked various dialogs shown by notepad and explorer/shell). On the topic of this PR:

1. Our buttons are 146px wide, while it looks that the native ones are 148px. This is probably unnoticeable and some native buttons are 146px too (but I think those are in the dialogs not using per-monitor DPI awareness, i.e. simply scaled up from the standard 73). The height is good (42px for all) and, again, nobody is probably going to notice this anyhow, but I just don't understand where does the difference come from.
2. In the single line text control the text is not centred vertically on both displays, but this is much more noticeable on the high DPI one: on the normal one, our top/bottom margins (easily seen if you select the text in the control) are 2/4px respectively. In the "new" (i.e. it's not an EDIT) search control in Explorer they're 2/3. However on high DPI monitor the native control uses 5/5, while ours one still uses 2/4px and 2px is really too small, i.e. the blue selection rectangle almost touches the border, which doesn't look great. Weirdly enough, the left margin does change, from 5 to 8px. Reassuringly, we're not the only ones having problems with this: "Go To Line" dialog in notebook looks extremely ugly in high DPI as they somehow end up with 2/15px margins.

Anyhow, I'm just noting this down to share the results of my testing, it doesn't prevent me from merging this PR, thanks again for all the fixes!
